src/MongoDB.EntityFrameworkCore/Serializers/DictionarySerializationProvider.cs

```diff
@@ -0,0 +1,87 @@
+/* Copyright 2023-present MongoDB Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Reflection;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using MongoDB.Bson.Serialization;
+using MongoDB.Bson.Serialization.Serializers;
+
+namespace MongoDB.EntityFrameworkCore.Serializers;
+
+internal class DictionarySerializationProvider : BsonSerializationProviderBase
+{
+    private static readonly Type[] SupportedDictionaryTypes =
+        [typeof(Dictionary<,>), typeof(IDictionary<,>), typeof(IReadOnlyDictionary<,>), typeof(ReadOnlyDictionary<,>)];
+
+    public static readonly DictionarySerializationProvider Instance = new();
+
+    public static bool Supports(Type type)
+        => type.IsGenericType && SupportedDictionaryTypes.Contains(type.GetGenericTypeDefinition());
+
+    public override IBsonSerializer GetSerializer(Type type, IBsonSerializerRegistry serializerRegistry)
+    {
+        ArgumentNullException.ThrowIfNull(type);
+
+        if (type is {IsGenericType: true, ContainsGenericParameters: true})
+        {
+            throw new ArgumentException($"Generic type '{type.ShortDisplayName()}' has unassigned generic type parameters.",
```

Comment:
Should it says "Open generic types are not supported"?

Reply:
I went with this naming as the docs page for ContainsGenericParameters at https://learn.microsoft.com/en-us/dotnet/api/system.reflection.methodbase.containsgenericparameters?view=net-8.0 refers to them as

> true if the current MethodBase object represents a generic method that contains unassigned generic type parameters; otherwise, false.